src/storage/StorageHttpDownloadHandler.cpp

```diff
@@ -0,0 +1,171 @@
+    }
+
+    if (auto hadoopHome = std::getenv("HADOOP_HOME")) {
+        LOG(INFO) << "Hadoop Path : " << hadoopHome;
+        std::vector<std::string> parts;
+        folly::split(",", partitions_, parts, true);
+        if (downloadSSTFiles(hdfsUrl_, hdfsPort_, hdfsPath_, parts, localPath_)) {
+            ResponseBuilder(downstream_)
+                .status(200, "SSTFile download successfully")
+                .body("SSTFile download successfully")
+                .sendWithEOM();
+        } else {
+            ResponseBuilder(downstream_)
+                .status(404, "SSTFile download failed")
+                .sendWithEOM();
+        }
+    } else {
+        ResponseBuilder(downstream_)
+            .status(404, "HADOOP_HOME not exist")
+            .sendWithEOM();
+    }
+}
+
+
+void StorageHttpDownloadHandler::onUpgrade(UpgradeProtocol) noexcept {
+    // Do nothing
+}
+
+
+void StorageHttpDownloadHandler::requestComplete() noexcept {
+    delete this;
+}
+
+
+void StorageHttpDownloadHandler::onError(ProxygenError error) noexcept {
+    LOG(ERROR) << "Web service StorageHttpDownloadHandler got error: "
+               << proxygen::getErrorString(error);
+}
+
+std::string StorageHttpDownloadHandler::toStr(folly::dynamic& vals) const {
+    std::stringstream ss;
+    for (auto& counter : vals) {
+        auto& val = counter["value"];
+        ss << counter["name"].asString() << "="
+           << val.asString()
+           << "\n";
+    }
+    return ss.str();
+}
+
+bool StorageHttpDownloadHandler::downloadSSTFiles(const std::string& url,
+                                          int port,
+                                          const std::string& path,
+                                          const std::vector<std::string>& parts,
+                                          const std::string& local) {
+    static folly::IOThreadPoolExecutor executor(FLAGS_download_thread_num);
+    std::condition_variable cv;
+    std::mutex lock;
+    std::atomic<int> completed(0);
+    std::atomic<bool> successful{true};
+
+    for (auto& part : parts) {
+        auto downloader = [&]() {
+            auto remotePath = folly::stringPrintf("hdfs://%s:%d%s/%d",
+                                                  url.c_str(), port, path.c_str(),
+                                                  atoi(part.c_str()) + 1);
+            LOG(INFO) << "File Path " << remotePath;
+            auto command = folly::stringPrintf("hdfs dfs -copyToLocal %s %s",
+                                               remotePath.c_str(), local.c_str());
+            LOG(INFO) << "Download SST Files: " << command;
+            auto result = ProcessUtils::runCommand(command.c_str());
+            if (!result.ok()) {
+                LOG(ERROR) << "Failed to download SST Files: " << remotePath;
+                successful = false;
+            }
+            std::unique_lock<std::mutex> uniqueLock(lock);
+            completed++;
+            cv.notify_one();
```

Comment:
why not notify the other thread after completed == parts.size()? otherwise cause useless thread schedule

Reply:
Good Catch  & fixed .